python/ray/serve/_private/proxy_router.py

```diff
@@ -108,17 +108,19 @@ def update_routes(self, endpoints: Dict[DeploymentID, EndpointInfo]) -> None:
             if endpoint in self.handles:
                 existing_handles.remove(endpoint)
             else:
-                handle = self._get_handle(endpoint.name, endpoint.app_name).options(
-                    # Streaming codepath isn't supported for Java.
-                    stream=not info.app_is_cross_language,
-                    _prefer_local_routing=RAY_SERVE_PROXY_PREFER_LOCAL_NODE_ROUTING,
-                    _source=DeploymentHandleSource.PROXY,
-                )
+                handle = self._get_handle(endpoint.name, endpoint.app_name)
+                # NOTE(zcin): since the router is eagerly initialized here,
+                # it will receive the replica set from the controller early.
+                if not handle.is_initialized:
```

Comment:
why do we need this check? shouldn't it be an assertion that the handle is _not_ initialized yet?

otherwise these important options will not be set..

Reply:
hmm yeah it's because `get_handle` will return a cached version of the handle, and this can happen because this code is executed every time the route table is updated.

Reply:
I see -- should we just explicitly keep a dictionary cache of handles here instead of relying on `get_handle` caching implicitly?

Reply:
in general the `get_handle` caching is going to be somewhat error prone with the `init` behavior

Reply:
yeah agreed ... took a look at the code, I think I can consolidate the proxy routers so that there is just a single "cache" of handles, right now there's two caches because there's two separate proxy router classes. will address this in a followup to keep the changes isolated.

As for how `get_handle` caching interacts with `init`, I think we might need a better story. In theory, the user is not expected to want to initialize same deployment handle twice with different set of options, but it is pretty confusing how the secret caching behavior can return an already-initialized handle.